### PR TITLE
Card Stack MVP implementation

### PR DESCRIPTION
**Pull request by bfollington**
https://github.com/subconsciousnetwork/subconscious/assets/5009316/da7105a9-5fb9-4a7c-a7a5-c4440dd5b470

https://github.com/subconsciousnetwork/subconscious/assets/5009316/c520736d-3ea4-4e32-844c-fd9b335e7e52

# Changes

- Add Deck tab
- Developer settings for Feed and Deck tab
- Implement MVP card drawing exploration concept

### CODE REVIEW

File: xcode/Subconscious/Shared/Components/Deck/DeckNavigationView.swift [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    var body: some View {
+        DetailStackView(app: app, store: detailStack) {
+            VStack {
+                VStack(alignment: .leading) {
+                    HStack(alignment: .center, spacing: AppTheme.unit3) {
+                        if case let .entry(_, author, _) = store.state.topCard?.card,
+                           let name = author.toNameVariant() {
+                            ProfilePic(pfp: author.pfp, size: .large)
+                            
+                            VStack(alignment: .leading, spacing: AppTheme.unit) {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Deck/DeckNavigationView.swift:33]: Remove wrapping VStack

File: xcode/Subconscious/Shared/Components/Deck/DeckNavigationView.swift [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                            store.send(.cardPickedUp)
+                        },
+                        onSwipeAbandoned: {
+                            store.send(.cardReleased)
+                        },
+                        onCardTapped: { card in
+                            store.send(.cardTapped(card))
+                        }
+                    )
+                    .offset(x: 0, y: -AppTheme.unit * 8)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Deck/DeckNavigationView.swift:74]: Frustrating manual adjustment to position the card stack in the right spot

File: xcode/Subconscious/Shared/Components/Deck/DeckView.swift [hunk 3]
```diff
@@ -0,0 +1,638 @@
...
+        .onAppear {
+            store.send(.appear)
+        }
+        .frame(maxWidth: .infinity)
+        /// Replay some app actions on feed store
+        .onReceive(
+            app.actions.compactMap(DeckAction.from),
+            perform: store.send
+        )
+        /// Replay some feed actions on app store
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Deck/DeckView.swift:118]: Update comment

File: xcode/Subconscious/Shared/Library/MathUtilities.swift [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+//  MathUtilities.swift
+//  Subconscious
+//
+//  Created by Ben Follington on 16/11/2023.
+//
+
+import Foundation
+
+extension Comparable {
+    public static func clamp<T: Comparable>(x: T, min lower: T, max upper: T) -> T {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Library/MathUtilities.swift:11]: Consider a test?

File: xcode/Subconscious/Shared/Services/DatabaseService.swift [hunk 5]
```diff
@@ -1171,6 +1171,48 @@ final class DatabaseService {
         })
         .first
     }
+    
+    func readRandomFreshEntryForCard(owner: Did?, seen: [Slashlink]) -> EntryStub? {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/DatabaseService.swift:1175]: Rename: "unseen" entry

File: xcode/Subconscious/Shared/Services/UserProfileService.swift [hunk 6]
```diff
@@ -194,7 +194,7 @@ actor UserProfileService {
     
     /// Attempt to read & deserialize a user `_profile_.json` from the indexed copy.
     /// This can only work for user's we follow & have indexed, otherwise it returns nil
-    func readProfileFromDb(
+    public func readProfileFromDb(
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Services/UserProfileService.swift:197]: Revert

File: xcode/Subconscious/Shared/Components/Deck/CardStackView.swift [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+            GeometryReader { geo in
+                ZStack {
+                    let deck = Array(deck.enumerated())
+                    ForEach(deck, id: \.element.id) { index, card in
+                        if (index >= current - 1 && index < current + 3) {
+                            VStack {
+                                Spacer()
+                                
+                                let stackFactor = stackFactor(for: index)
+                                CardView(entry: card)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Deck/CardStackView.swift:149]: I plan to break this out into modular chunks eventually, but while we're still churning on styles I think it's best to leave it as-is.

File: xcode/Subconscious/Shared/Components/Deck/CardModel.swift [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    case entry(
+        entry: EntryStub,
+        author: UserProfile,
+        backlinks: [EntryStub]
+    )
+    case action(String)
+}
+
+struct CardModel: Identifiable, Equatable, Hashable {
+    var id: UUID = UUID()
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Deck/CardModel.swift:21]: Allows us to repeat cards in the deck but still uniquely identify them

File: xcode/Subconscious/Shared/Components/Deck/CardStackView.swift [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+            case .action(let string):
+                // TEMP
+                VStack {
+                    Image(systemName: "scribble.variable")
+                        .resizable()
+                        .aspectRatio(contentMode: .fit)
+                        .frame(width: 64, height: 64)
+                    Text(string)
+                }
+                .foregroundStyle(.secondary)
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Deck/CardStackView.swift:59]: We currently have no meaningful actions, but I expect they're coming soon.

File: xcode/Subconscious/Shared/Components/Deck/DeckNavigationView.swift [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    var detailStack: ViewStore<DetailStackModel> {
+        store.viewStore(
+            get: DeckDetailStackCursor.get,
+            tag: DeckDetailStackCursor.tag
+        )
+    }
+    
+    var body: some View {
+        DetailStackView(app: app, store: detailStack) {
+            VStack {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Deck/DeckNavigationView.swift:26]: Superfluous?

File: xcode/Subconscious/Shared/Components/Deck/DeckTheme.swift [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+//  Subconscious (iOS)
+//
+//  Created by Ben Follington on 22/11/2023.
+//
+
+import SwiftUI
+
+enum DeckTheme {}
+
+extension DeckTheme {
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Deck/DeckTheme.swift:12]: Collected all my constants here

File: xcode/Subconscious/Shared/Components/Deck/DeckView.swift [hunk 12]
```diff
@@ -0,0 +1,554 @@
...
+}
+
+extension DeckAction {
+    static func from(_ action: AppAction) -> Self? {
+        switch action {
+        default:
+            return nil
+        }
+    }
+}
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Deck/DeckView.swift:126]: Placeholder, we'll need these to wire up notifications.

File: xcode/Subconscious/Shared/Components/Deck/DeckView.swift [hunk 13]
```diff
@@ -0,0 +1,554 @@
...
+    var deck: [CardModel] = []
+    
+    // The set of cards to avoid drawing again, if possible
+    var seen: Set<EntryStub> = []
+    var pointer: Int = 0
+    var author: UserProfile? = nil
+    var selectionFeedback = UISelectionFeedbackGenerator()
+    var feedback = UIImpactFeedbackGenerator()
+    
+    static func insertAtRandomIndex<T>(
```
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Deck/DeckView.swift:209]: Consider adding test, relocating function

File: xcode/Subconscious/Shared/Components/Deck/DeckView.swift [hunk 14]
```diff
@@ -0,0 +1,575 @@
...
+    /// Search HUD
+    var isSearchPresented = false
+    var search = SearchModel(
+        placeholder: "Search or create..."
+    )
+    
+    var deck: Array<CardModel> = []
+    
+    // The set of cards to avoid drawing again, if possible
+    var seen: Set<EntryStub> = []
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Deck/DeckView.swift:201]: 😌 nice

File: xcode/Subconscious/Shared/Components/Deck/DeckView.swift [hunk 15]
```diff
@@ -0,0 +1,575 @@
...
+    var pointer: Int = 0
+    var author: UserProfile? = nil
+    var selectionFeedback = UISelectionFeedbackGenerator()
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Deck/DeckView.swift:204]: This should be part of the environment. We want models to be dumb data only.

Non-blocking, but let's address in a follow-up.

File: xcode/Subconscious/Shared/Components/Deck/DeckView.swift [hunk 16]
```diff
@@ -0,0 +1,575 @@
...
+    var feedback = UIImpactFeedbackGenerator()
```
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Deck/DeckView.swift:205]: This should be part of the environment. We want models to be dumb data only.

Non-blocking, but let's address in a follow-up.

**Review by gordonbrander — APPROVED** [review]
Looks great to me.

**Pull request merged**